mirgecom/tag_cells.py

```diff
@@ -0,0 +1,159 @@
+            & s_0-\kappa \le s_e \le s_0+\kappa \\
+            1, & s_e > s_0+\kappa
+        \end{cases}
+
+where:
+- $\varepsilon_e$ is the element viscosity
+- $\varepsilon_0 ~\sim h/p$ is a reference viscosity
+- $s_e = \log_{10}{S_e} \sim 1/p^4$ is the smoothness indicator
+- $s_0$ is a reference smoothness value
+- $\kappa$ controls the width of the transition between 0 to $\varepsilon_0$
+
+Smoothness Indicator Evaluation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. autofunction:: smoothness_indicator
+"""
+
+__copyright__ = """
+Copyright (C) 2020 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import loopy as lp
+from meshmode.dof_array import DOFArray
+from modepy import vandermonde
+
+
+def linear_operator_kernel():
+    """Apply linear operator to all elements."""
+    from meshmode.array_context import make_loopy_program
+
+    knl = make_loopy_program(
+        """{[iel,idof,j]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in}""",
+        "result[iel,idof] = sum(j, mat[idof, j] * vec[iel, j])",
+        name="modal_decomp",
+    )
+    knl = lp.tag_array_axes(knl, "mat", "stride:auto,stride:auto")
+    return knl
+
+
+def compute_smoothness_indicator():
+    """Compute the smoothness indicator for all elements."""
+    from meshmode.array_context import make_loopy_program
+
+    knl = make_loopy_program(
+        """{[iel,idof,j,k]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in and
+        0<=k<ndiscr_nodes_in}""",
+        "result[iel,idof] = "
+        "sum(k,vec[iel,k]*vec[iel,k]*modes[k])/sum(j,"
+        " vec[iel,j]*vec[iel,j]+1.0e-12/ndiscr_nodes_in)",
+        name="smooth_comp",
+    )
+    return knl
+
+
+def smoothness_indicator(u, discr, kappa=1.0, s0=-6.0):
+    """Calculate the smoothness indicator."""
```

Comment:
Should have a documented interface.

Reply:
Added interface parameter documentation: https://github.com/illinois-ceesd/mirgecom/pull/213/commits/218cb523e501a27c1aabb6ec56899efb9975a816